### ISSUE TITLE
Notifier parameter handling

### ISSUE THREAD

**Issue by scottcwilson** [zencart/documentation#216]
I thought it was incredibly clever how `EditOrdersAdminObserver.php` uses generic names for the formal parameters of the `update` function, and then describes them in comments for each event type. 

Would you please update 

https://docs.zen-cart.com/dev/code/notifiers/ 

and describe this technique?  

(While you're there, it would be great to give that article a scrub and improve the examples.) 

Thank you!

**Comment by drbyte** [comment]
NO!!!!

The use of generic names makes following the bouncing ball particularly difficult when trying to read and understand and debug the code.

**Comment by scottcwilson** [comment]
What's your recommendation for handling event types with different parameter lists then?

**Comment by drbyte** [comment]
Simply use the calculated `updateNotifierNameCamelCased()` functions.

If a notifier name is `NOTIFY_FOO_BAZ` and you provide a function named `updateNotifyFooBaz` which accepts all the same parameters as the generic `update` function, then it will auto-resolve the correct method name to use, and directly interact with the supplied variables.

**Comment by lat9** [comment]
FWIW, I _really_, **really**, don't like that **_CamelCasingFunctionNames_**.  It, IMO, makes it very difficult to find what processing is acting on a given notification.

**Comment by scottcwilson** [comment]
Ditto.  It's OBVIOUS what's going on in Cindy's logic; what you're describing is black magic.  The only mods I see that use this technique are order delivery date zen Non CAPTCHA.  I didn't know about this until just this moment.

**Comment by drbyte** [comment]
The file-download infrastructure uses it extensively: 
https://github.com/zencart/zencart/blob/v156/includes/classes/observers/auto.downloads_via_url.php

**Comment by drbyte** [comment]
It was added in v1.5.2 when the additional mutable parameters feature was added.

**Comment by scottcwilson** [comment]
> The file-download infrastructure uses it extensively

This means "the authors of the code think it's awesome!"  Adoption by modders would mean others agree - quite a different thing.

**Comment by drbyte** [comment]
> It's OBVIOUS what's going on in Cindy's logic

Only obvious insomuch as you've written a book of comments that you have to reference and translate, and maintain. I find the generic variable names highly unfriendly. Plus it requires adding nested `switch`/etc loops, which are also hard to follow because you never know if there's more code to be run or other things that might be getting done elsewhere.

Using named methods (functions) keeps context, and the named parameters supplies direct meaning ... which means it's self-documenting. This benefits by making human reading easier, maintenance simpler, and less likely to forget which parameter is which because it's actually the real parameter, not some abstract generic number.

That said, clearly you both have a different perspective. What exactly do you find difficult about the convention?
Is the issue that the name of the notifier hook is not searchable because the underscores have been removed to follow the camel-casing convention? 
If so, that's an ideal case for a single-line comment for the method, where you provide the name of the notifier hook. That way your code editor can find it ... if that's what you find difficult about the convention.
Or we could consider supporting snake_cased naming if that would be easier.

**I'm open to usability discussion about it.**

**Comment by scottcwilson** [comment]
I'm going to split out the other (hopefully non-controversial) improvements to this page to a separate ticket.

**Comment by lat9** [comment]
> 
> 
> > It's OBVIOUS what's going on in Cindy's logic
> 
> Only obvious insomuch as you've written a book of comments that you have to reference and translate, and maintain. I find the generic variable names highly unfriendly. Plus it requires adding nested `switch`/etc loops, which are also hard to follow because you never know if there's more code to be run or other things that might be getting done elsewhere.
> 
> Using named methods (functions) keeps context, and the named parameters supplies direct meaning ... which means it's self-documenting. This benefits by making human reading easier, maintenance simpler, and less likely to forget which parameter is which because it's actually the real parameter, not some abstract generic number.
> 
> That said, clearly you both have a different perspective. What exactly do you find difficult about the convention?
> Is the issue that the name of the notifier hook is not searchable because the underscores have been removed to follow the camel-casing convention?
> If so, that's an ideal case for a single-line comment for the method, where you provide the name of the notifier hook. That way your code editor can find it ... if that's what you find difficult about the convention.
> Or we could consider supporting snake_cased naming if that would be easier.
> 
> **I'm open to usability discussion about it.**

Many times, when I'm freshly viewing a new client's file-system, I'm seeing various observer classes and wondering if there's going to be a conflict (i.e. two observers on a single notification).  I'll take the notification _event_ and look it up using the Developer's Tool Kit.  For the camel-cased construct, I'll see the observers' `attach` to the event, but don't see where the processing is actually performed.

Essentially, to do that, I have to (in my head) camelCase the notification event to search for that format (lost time and client's money).

The reason I went to the 'document the expected parameters' type comments was that at least once in the past, the notification parameters were _changed_ from one Zen Cart release to the other, so I prefer to document what I've designed to and keep the actual `update` processing to a minimum.  Normally (noting that there are _always_ exceptions to the rule), if the processing is one or two lines I keep the use of the `$p[1-9]` parameters since the comments are "right there".  If there's a more complex bit of processing to perform, I'll create a `protected` or `private` method to deal with the changes and use the more obvious name of the actual variable.

I'm also not a big fan of the overly-long function names that result, for example (from the orders.php class) `NOTIFY_ORDER_CART_EXTERNAL_TAX_DURING_ORDER_CREATE` turns into
`updateNotifyOrderCartExternalTaxDuringOrderCreate` which is a 'mouthful' of a function name and doesn't necessarily add any value to the code's readability.

**Comment by scottcwilson** [comment]
> ... at least once in the past, the notification parameters were changed from one Zen Cart release to the other

Please for the love let's never have this happen again.  This happened to me too and it was a huge pain to debug.

**Comment by drbyte** [comment]
@zcwilt if we modify the **camelCase** to also look for **snake_case** then searches using the Developers Toolkit or an IDE would find the uses of those notifier event names.

Of course, retaining backward support for camelCase would be necessary, but we could lookup as snake_case first.

Thoughts?

**Comment by mc12345678** [comment]
> @zcwilt if we modify the camelCase to also look for snake_case then searches using the Developers Toolkit or an IDE would find the uses of those notifier event names.
>
> Of course, retaining backward support for camelCase would be necessary, but we could lookup as snake_case first.
>
> Thoughts?

I'm 100% behind this and have been since camelcase was [first proposed in the zencart "branch"](https://github.com/zencart/zencart/issues/312#issuecomment-171637363).

Now that said and I'm obviously late to the discussion, but I've been using the `updateCamelCased` code for any observer software that I write, but of course I also observe backwards compatibility much more than either of you two.  I had begun initially with just putting the notifier name, but then at a point in time I began placing a note above the camelcased function that identifies the actual notifier as developed to support a ZC version so that such "changes" (See [this code](https://github.com/mc12345678/Stock_By_Attributes_Combined/blob/ad954a5adfceb187cabe507f26d9c3a63192c406/includes/classes/observers/class.products_with_attributes_stock.php#L836-L841)), however small, can be immediately seen in newer versions... If such a change has an impact, then a determination can be made to drop previous compatibility as the receiving parameter may need some level of parsing (wasn't an array before, but now is, etc...).  This way, the search actually pulls up the code that is associated.  

Further, and though I haven't been 100% successful, I try to keep my `method`s in order with that identified in the attach area of the `__construct`.  That's more for "readability"... To then take it even further, I will include a fallback `update` method where (possibly) every parameter that is not supported by older ZC versions has a "default" value assigned (to prevent completely shutting down that site if the file is set to load) and yes, in that a "generic" parameter is used... When a value is *not* available through such a notifier (older version) but *is* available by global or some other method, then that other method is used to bring the variable into the method and the updateCamelCased version is called with all of the necessary parameters... (See [this code example](https://github.com/mc12345678/Stock_By_Attributes_Combined/blob/ad954a5adfceb187cabe507f26d9c3a63192c406/includes/classes/observers/class.products_with_attributes_stock.php#L1428-L1432)) To go even further, unfortunately in support of older Zen Cart versions, a modification may be necessary to the original file to provide some level of data that can not otherwise be retrieved from the default older file version... But I will say this last aspect is becoming a little less of a consideration for pre-1.5.3 Zen Cart software.

Anyways, I have/had begun a [modified version of the documentation on the notifier/observer class](https://github.com/mc12345678/documentation/commit/56f7d807b4d7b7dea1f6218d19e96d8dd6ce4e20) to take into consideration *other* ways/considerations to adapt the use of the observer class.

**Comment by scottcwilson** [comment]
How do we move forward on this issue?  It feels like we're stuck.

**Comment by drbyte** [comment]
Starting with 1.5.7 the `update` method has 2 alternate ways of being called, for a total of 3 approaches supported.
I prefer using the snake-case, but of course this only works since 1.5.7.
They fire the first one found matching the following pattern:
- `snake_case` (lowercase) of attached event name if starts with `NOTIFY_` or `NOTIFIER_`, eg: `notify_cart_something()`
- `camelCase`: uses `update` prefix followed by camelcased event name, eg: `updateNotifyCartSomething()`
- generic `update` method which fires if neither of the others exist. Must check the `$eventID` to know what event was actually fired.

All 3 approaches use the same parameter signature.

**Comment by scottcwilson** [comment]
Thanks @drbyte for updating this help page:
https://docs.zen-cart.com/dev/code/notifiers/#event-specific-update-methods
